File: packages/oxlint-plugin-react-doctor/src/plugin/utils/can-content-editable-be-tabbable.ts
import { HTML_TAGS } from "../constants/html-tags.js";
import type { ScopeAnalysis } from "../semantic/scope-analysis.js";
import type { EsTreeNode } from "./es-tree-node.js";
import type { EsTreeNodeOfType } from "./es-tree-node-of-type.js";
import { getDirectConstInitializer } from "./get-direct-const-initializer.js";
import { getElementType } from "./get-element-type.js";
import { hasJsxPropIgnoreCase } from "./has-jsx-prop-ignore-case.js";
import { isNodeOfType } from "./is-node-of-type.js";
import { stripParenExpression } from "./strip-paren-expression.js";

interface ContentEditablePossibilities {
  canBeDisabled: boolean;
  canBeEnabled: boolean;
}

const ENABLED_CONTENT_EDITABLE_VALUES: ReadonlySet<string> = new Set([
  "",
  "plaintext-only",
  "true",
]);

const ENABLED_CONTENT_EDITABLE: ContentEditablePossibilities = {
  canBeDisabled: false,
  canBeEnabled: true,
};

const DISABLED_CONTENT_EDITABLE: ContentEditablePossibilities = {
  canBeDisabled: true,
  canBeEnabled: false,
};

const UNKNOWN_CONTENT_EDITABLE: ContentEditablePossibilities = {
  canBeDisabled: true,
  canBeEnabled: true,
};

const mergeContentEditablePossibilities = (
  left: ContentEditablePossibilities,
  right: ContentEditablePossibilities,
): ContentEditablePossibilities => ({
  canBeDisabled: left.canBeDisabled || right.canBeDisabled,
  canBeEnabled: left.canBeEnabled || right.canBeEnabled,
});

const resolveContentEditableExpression = (
  rawExpression: EsTreeNode,
  scopes: ScopeAnalysis,
  visitedSymbolIds: Set<number>,
): ContentEditablePossibilities => {
  const expression = stripParenExpression(rawExpression);
  if (isNodeOfType(expression, "Literal")) {
    if (typeof expression.value === "boolean") {
      return expression.value ? ENABLED_CONTENT_EDITABLE : DISABLED_CONTENT_EDITABLE;
    }
    if (typeof expression.value === "string") {
      return ENABLED_CONTENT_EDITABLE_VALUES.has(expression.value.toLowerCase())
        ? ENABLED_CONTENT_EDITABLE
        : DISABLED_CONTENT_EDITABLE;
    }
    return DISABLED_CONTENT_EDITABLE;
  }
  if (isNodeOfType(expression, "ConditionalExpression")) {
    return mergeContentEditablePossibilities(
      resolveContentEditableExpression(expression.consequent, scopes, visitedSymbolIds),
      resolveContentEditableExpression(expression.alternate, scopes, visitedSymbolIds),
    );
  }
  if (isNodeOfType(expression, "Identifier")) {
    const symbol = scopes.referenceFor(expression)?.resolvedSymbol;
    if (!symbol || visitedSymbolIds.has(symbol.id)) return UNKNOWN_CONTENT_EDITABLE;
    const initializer = getDirectConstInitializer(symbol);
    if (!initializer) return UNKNOWN_CONTENT_EDITABLE;
    const nextVisitedSymbolIds = new Set(visitedSymbolIds);
    nextVisitedSymbolIds.add(symbol.id);
    return resolveContentEditableExpression(initializer, scopes, nextVisitedSymbolIds);
  }
  return UNKNOWN_CONTENT_EDITABLE;
};

const getContentEditablePossibilities = (
  node: EsTreeNodeOfType<"JSXOpeningElement">,
  scopes: ScopeAnalysis,
): ContentEditablePossibilities | null => {
  const attribute = hasJsxPropIgnoreCase(node.attributes, "contenteditable");
  if (!attribute) return null;
  if (!attribute.value) return ENABLED_CONTENT_EDITABLE;
  if (isNodeOfType(attribute.value, "Literal")) {
    return resolveContentEditableExpression(attribute.value, scopes, new Set());
  }
  if (isNodeOfType(attribute.value, "JSXExpressionContainer")) {
    return resolveContentEditableExpression(attribute.value.expression, scopes, new Set());
  }
  return UNKNOWN_CONTENT_EDITABLE;
};

const hasDefinitelyEnabledContentEditableAncestor = (
  node: EsTreeNodeOfType<"JSXOpeningElement">,
  scopes: ScopeAnalysis,
  settings: Readonly<Record<string, unknown>> | undefined,
): boolean => {
  let ancestor = node.parent?.parent;
  while (ancestor) {
    if (isNodeOfType(ancestor, "JSXElement")) {
      const openingElement = ancestor.openingElement;
      if (HTML_TAGS.has(getElementType(openingElement, settings))) {
        const possibilities = getContentEditablePossibilities(openingElement, scopes);
        if (possibilities?.canBeEnabled && !possibilities.canBeDisabled) return true;
      }
    }
    ancestor = ancestor.parent;
  }
  return false;
};

export const canContentEditableBeTabbable = (
  node: EsTreeNodeOfType<"JSXOpeningElement">,
  scopes: ScopeAnalysis,
  settings: Readonly<Record<string, unknown>> | undefined,
): boolean => {
  const possibilities = getContentEditablePossibilities(node, scopes);
  if (!possibilities?.canBeEnabled) return false;
  return !hasDefinitelyEnabledContentEditableAncestor(node, scopes, settings);
};

File: native/oxlint/rules/can-content-editable-be-tabbable.rs
#[derive(Clone, Copy)]
struct ContentEditablePossibilities {
    can_be_disabled: bool,
    can_be_enabled: bool,
    can_be_inherited: bool,
}

const ENABLED_CONTENT_EDITABLE: ContentEditablePossibilities = ContentEditablePossibilities {
    can_be_disabled: false,
    can_be_enabled: true,
    can_be_inherited: false,
};
const DISABLED_CONTENT_EDITABLE: ContentEditablePossibilities = ContentEditablePossibilities {
    can_be_disabled: true,
    can_be_enabled: false,
    can_be_inherited: false,
};
const INHERITED_CONTENT_EDITABLE: ContentEditablePossibilities = ContentEditablePossibilities {
    can_be_disabled: false,
    can_be_enabled: false,
    can_be_inherited: true,
};
const UNKNOWN_CONTENT_EDITABLE: ContentEditablePossibilities = ContentEditablePossibilities {
    can_be_disabled: true,
    can_be_enabled: true,
    can_be_inherited: true,
};

fn can_content_editable_be_tabbable<'a>(
    node: &crate::AstNode<'a>,
    opening_element: &'a oxc_ast::ast::JSXOpeningElement<'a>,
    ctx: &crate::context::LintContext<'a>,
) -> bool {
    let Some(possibilities) = get_content_editable_possibilities(opening_element, ctx) else {
        return false;
    };
    possibilities.can_be_enabled && !has_definitely_enabled_content_editable_ancestor(node, ctx)
}

fn has_definitely_enabled_content_editable_ancestor<'a>(
    node: &crate::AstNode<'a>,
    ctx: &crate::context::LintContext<'a>,
) -> bool {
    let mut did_skip_current_element = false;
    for ancestor in ctx.nodes().ancestors(node.id()) {
        let oxc_ast::AstKind::JSXElement(element) = ancestor.kind() else {
            continue;
        };
        if !did_skip_current_element {
            did_skip_current_element = true;
            continue;
        }
        if !crate::globals::HTML_TAG.contains(
            crate::utils::get_element_type(ctx, &element.opening_element).as_ref(),
        ) {
            return false;
        }
        let Some(possibilities) =
            get_content_editable_possibilities(&element.opening_element, ctx)
        else {
            continue;
        };
        if possibilities.can_be_enabled
            && !possibilities.can_be_disabled
            && !possibilities.can_be_inherited
        {
            return true;
        }
        if !(possibilities.can_be_inherited
            && !possibilities.can_be_disabled
            && !possibilities.can_be_enabled)
        {
            return false;
        }
    }
    false
}

fn get_content_editable_possibilities<'a>(
    opening_element: &'a oxc_ast::ast::JSXOpeningElement<'a>,
    ctx: &crate::context::LintContext<'a>,
) -> Option<ContentEditablePossibilities> {
    let oxc_ast::ast::JSXAttributeItem::Attribute(attribute) =
        crate::utils::has_jsx_prop_ignore_case(opening_element, "contenteditable")?
    else {
        return Some(UNKNOWN_CONTENT_EDITABLE);
    };
    let Some(value) = attribute.value.as_ref() else {
        return Some(ENABLED_CONTENT_EDITABLE);
    };
    Some(match value {
        oxc_ast::ast::JSXAttributeValue::StringLiteral(string_literal) => {
            content_editable_string_value(string_literal.value.as_str())
        }
        oxc_ast::ast::JSXAttributeValue::ExpressionContainer(container) => container
            .expression
            .as_expression()
            .map_or(UNKNOWN_CONTENT_EDITABLE, |expression| {
                resolve_content_editable_expression(expression, ctx, &mut Vec::new())
            }),
        _ => UNKNOWN_CONTENT_EDITABLE,
    })
}

fn resolve_content_editable_expression<'a>(
    expression: &'a oxc_ast::ast::Expression<'a>,
    ctx: &crate::context::LintContext<'a>,
    visited_symbol_ids: &mut Vec<oxc_semantic::SymbolId>,
) -> ContentEditablePossibilities {
    use oxc_ast::ast::Expression;

    match expression.get_inner_expression() {
        Expression::BooleanLiteral(boolean_literal) => {
            if boolean_literal.value {
                ENABLED_CONTENT_EDITABLE
            } else {
                DISABLED_CONTENT_EDITABLE
            }
        }
        Expression::StringLiteral(string_literal) => {
            content_editable_string_value(string_literal.value.as_str())
        }
        Expression::ConditionalExpression(conditional_expression) => merge_content_editable(
            resolve_content_editable_expression(
                &conditional_expression.consequent,
                ctx,
                &mut visited_symbol_ids.clone(),
            ),
            resolve_content_editable_expression(
                &conditional_expression.alternate,
                ctx,
                &mut visited_symbol_ids.clone(),
            ),
        ),
        Expression::Identifier(identifier) => {
            let Some(symbol_id) = ctx
                .scoping()
                .get_reference(identifier.reference_id())
                .symbol_id()
            else {
                return UNKNOWN_CONTENT_EDITABLE;
            };
            if visited_symbol_ids.contains(&symbol_id) {
                return UNKNOWN_CONTENT_EDITABLE;
            }
            let declaration = ctx.symbol_declaration(symbol_id);
            let oxc_ast::AstKind::VariableDeclarator(declarator) = declaration.kind() else {
                return UNKNOWN_CONTENT_EDITABLE;
            };
            let parent = ctx.nodes().parent_node(declaration.id());
            let oxc_ast::AstKind::VariableDeclaration(variable_declaration) = parent.kind() else {
                return UNKNOWN_CONTENT_EDITABLE;
            };
            if !variable_declaration.kind.is_const()
                || declarator
                    .id
                    .get_binding_identifier()
                    .is_none_or(|binding_identifier| binding_identifier.symbol_id() != symbol_id)
            {
                return UNKNOWN_CONTENT_EDITABLE;
            }
            let Some(initializer) = declarator.init.as_ref() else {
                return UNKNOWN_CONTENT_EDITABLE;
            };
            visited_symbol_ids.push(symbol_id);
            let possibilities =
                resolve_content_editable_expression(initializer, ctx, visited_symbol_ids);
            visited_symbol_ids.pop();
            possibilities
        }
        Expression::NullLiteral(_)
        | Expression::NumericLiteral(_)
        | Expression::BigIntLiteral(_)
        | Expression::RegExpLiteral(_) => INHERITED_CONTENT_EDITABLE,
        _ => UNKNOWN_CONTENT_EDITABLE,
    }
}

fn content_editable_string_value(value: &str) -> ContentEditablePossibilities {
    match value.to_ascii_lowercase().as_str() {
        "" | "plaintext-only" | "true" => ENABLED_CONTENT_EDITABLE,
        "false" => DISABLED_CONTENT_EDITABLE,
        _ => INHERITED_CONTENT_EDITABLE,
    }
}

fn merge_content_editable(
    left: ContentEditablePossibilities,
    right: ContentEditablePossibilities,
) -> ContentEditablePossibilities {
    ContentEditablePossibilities {
        can_be_disabled: left.can_be_disabled || right.can_be_disabled,
        can_be_enabled: left.can_be_enabled || right.can_be_enabled,
        can_be_inherited: left.can_be_inherited || right.can_be_inherited,
    }
}
